### ISSUE TITLE
General stability issues

### ISSUE THREAD

**Issue by clayne11** [gadicc/meteor-hmr#63]
I'm having issues with my app randomly crashing.

It's happening pretty frequently. I'll post a stack next time it happens  but it doesn't look like a consistent error. I'm wondering if it's related to #34 and #62.

**Comment by clayne11** [comment]
Here's a stack:

```

/Users/clayne/.meteor/packages/less/.2.6.0.cdglvi++os+web.browser+web.cordova/plugin.compileLessBatch.os/npm/node_modules/meteor/promise/node_modules/meteor-promise/promise_server.js:116
      throw error;
            ^
Error: UNKNOWN, readdir '/Users/clayne/workspace/r3ntr_meteor/public/font/champagne'

[gadicc:hot] Accelerator (cMz): disconnect event received, exiting.
```
The `Error: UNKNOWN, readdir` references a different file randomly every time this happens.

**Comment by gadicc** [comment]
Different file, but always an `UNKNOWN, readdir` error?

Did this only start when you switched to vim?  The less compiler might have the same issue with atomic writes that we do.  Although not sure why anything in `public/font/champagne` would be affected.  It's not a network mount or anything, right?  Also, can you include the full stack trace next time too?  This and some of the other errors are not coming from the hot code, but apparently don't happen without the hot code.

I haven't experienced any of them personally but they all seem related to the node file watching code which differs quite substantially by OS.  Something we're doing seems to be affecting other stuff and I'm going to try track this down.

**Comment by clayne11** [comment]
Updated to `refactor.9` and immediately got this issue again.

```
/Users/clayne/.meteor/packages/less/.2.6.0.cdglvi++os+web.browser+web.cordova/plugin.compileLessBatch.os/npm/node_modules/meteor/promise/node_modules/meteor-promise/promise_server.js:116
      throw error;
            ^
Error: UNKNOWN, readdir '/Users/clayne/workspace/r3ntr_meteor/node_modules/webdriverio/node_modules/css/lib/stringify'

[gadicc:hot] Accelerator (mqe): disconnect event received, exiting.
```

That's the entire s tack trace.

The app only seems to actually crash when I connect to it from a web browser. Once this issue start happening I have to restart my computer to fix the issue. I've tried killing all Meteor related processes with:

`ps aux | grep -ie meteor | awk '{print $2}' | xargs kill -9`

but the issue persists until I restart my computer.

You mentioned that you think this issue might be related to watching the `babelrc` files. Maybe we should just not watch them? They're not files that change very often, and you could just mention in the docs that you need to restart Meteor after changing them. I think we're seeing a LOT of hassle for very little benefit in this case.

**Comment by gadicc** [comment]
Will be back on this later/tomorrow.  It the error always from the `less` package?  Did you ever check if you don't experience it without `ecmascript-hot`?  (Yeah, I know it's painful to give it up for a day :)).  The babelrc watcher thing is just my best guess, but yeah, I'm happy to try this.  Didn't realize you had to reboot every time too!  That's super weird.

**Comment by clayne11** [comment]
It's not always from the less package. As far as I can tell it's randomly a different file/package every time - I actually don't even use `less`, I'm not sure why it's in my project at all. I don't directly import it.

It does not happen without `ecmascript-hot`. I switched to `ecmascript-babel` yesterday and didn't experience it at all.

**Comment by gadicc** [comment]
That's even more worrying.... because `ecmascript-babelrc` is where the babelrc file watching code is :D  I'll have to have another look at the code tomorrow to see if I can find anything else that could be causing this.

**Comment by gadicc** [comment]
In `refactor.10`, I use a release of `babel-compiler-babelrc` which doesn't watch your .babelrc files, to see if it helps the eventemitter warning... but it sounds like this issue is something different.

**Comment by clayne11** [comment]
Still seeing issues with `refactor.12`.

I've recorded each of the following crashes with `refactor.12`:

```Error: UNKNOWN, readdir '/Users/clayne/workspace/r3ntr_meteor/node_modules/webdriverio/node_modules/wdio-dot-reporter/coverage/lcov-report'

Error: UNKNOWN, readdir '/Users/clayne/workspace/r3ntr_meteor/node_modules/xolvio-ddp/node_modules/bl/test'

Error: UNKNOWN, readdir '/Users/clayne/workspace/r3ntr_meteor/public/images/viewListing'

undefined:0
Error: UNKNOWN, readdir '/Users/clayne/workspace/r3ntr_meteor/packages/casalova-default-email-notifications'


Error: UNKNOWN, readdir '/Users/clayne/workspace/r3ntr_meteor/node_modules/xolvio-ddp/node_modules/mime-db'

undefined:0

^
SyntaxError: Unexpected end of input
    at Object.parse (native)
    at WebSocket.<anonymous> (/Users/clayne/.meteor/packages/gadicc_ecmascript-hot/.1.3.2-refactor.7.bth709++os+web.browser+web.cordova/plugin.compile-ecmascript-hot.os/npm/node_modules/meteor/gadicc_hot-build/node_modules/meteor-hotload-accelerator/lib/accelerator.js:88:22)
    at WebSocket.emit (events.js:98:17)
    at Receiver.self._receiver.ontext (/Users/clayne/.meteor/packages/meteor-tool/.1.3.2_2.hsij42++os.osx.x86_64+web.browser+web.cordova/mt-os.osx.x86_64/dev_bundle/lib/node_modules/ws/lib/WebSocket.js:682:10)
    at Receiver.opcodes.1.finish (/Users/clayne/.meteor/packages/meteor-tool/.1.3.2_2.hsij42++os.osx.x86_64+web.browser+web.cordova/mt-os.osx.x86_64/dev_bundle/lib/node_modules/ws/lib/Receiver.js:391:14)
    at /Users/clayne/.meteor/packages/meteor-tool/.1.3.2_2.hsij42++os.osx.x86_64+web.browser+web.cordova/mt-os.osx.x86_64/dev_bundle/lib/node_modules/ws/lib/Receiver.js:372:33
    at Receiver.expectData (/Users/clayne/.meteor/packages/meteor-tool/.1.3.2_2.hsij42++os.osx.x86_64+web.browser+web.cordova/mt-os.osx.x86_64/dev_bundle/lib/node_modules/ws/lib/Receiver.js:147:5)
    at Receiver.expectHandler (/Users/clayne/.meteor/packages/meteor-tool/.1.3.2_2.hsij42++os.osx.x86_64+web.browser+web.cordova/mt-os.osx.x86_64/dev_bundle/lib/node_modules/ws/lib/Receiver.js:371:16)
    at Receiver.add (/Users/clayne/.meteor/packages/meteor-tool/.1.3.2_2.hsij42++os.osx.x86_64+web.browser+web.cordova/mt-os.osx.x86_64/dev_bundle/lib/node_modules/ws/lib/Receiver.js:87:24)
    at Socket.firstHandler (/Users/clayne/.meteor/packages/meteor-tool/.1.3.2_2.hsij42++os.osx.x86_64+web.browser+web.cordova/mt-os.osx.x86_64/dev_bundle/lib/node_modules/ws/lib/WebSocket.js:663:22)
```

The final one had an actual stack trace.

Additionally, when trying to run `git status` on my app's repo, the following happened, and then the app crashed immediately afterwards:

```
warning: unable to access 'imports/.gitignore': Too many open files in system
warning: unable to access 'node_modules/.gitignore': Too many open files in system
warning: unable to access 'packages/.gitignore': Too many open files in system
warning: unable to access 'private/.gitignore': Too many open files in system
warning: unable to access 'public/.gitignore': Too many open files in system
warning: unable to access 'server/.gitignore': Too many open files in system
warning: unable to access 'tests/.gitignore': Too many open files in system
On branch rent_payments
Your branch is ahead of 'origin/rent_payments' by 1 commit.
  (use "git push" to publish your local commits)
nothing to commit, working directory clean
git_prompt_info:2: too many open files in system: /dev/null
git_prompt_info:3: too many open files in system: /dev/null
git_prompt_info:4: too many open files in system: /dev/null
```

**Comment by clayne11** [comment]
I had to switch to `gadicc:ecmascript-babelrc`. I'm having absolutely no issues with it.

The stability of `ecmsacript-hot` has been big problem. I've had to restart my computer about 7 times today and I'm not really able to get work done.

**Comment by gadicc** [comment]
Hey, this was actually very helpful!

I think your open file limit is just a little small.  You can check with `ulimit -n` (per session) and `sysctl kern.maxfiles`.

I'm assuming you have quite a large number of `.js` files in your project, right?  I'm not sure how the file watching on OSX works, but I guess it's using file descriptors, and now we've just doubled what you were watching before (once for Meteor, once for the accelerator).  I think you could probably trigger this just by running Meteor twice, but, from within the same shell (you'd have to `meteor&` and load the 2nd one with a different `.meteor` directory...).

If you do have a chance to try the Meteor twice thing, it would be great to confirm, because in case this doesn't trigger the same problem, Meteor could be doing something smarter than us, but I'm not really sure what else they could be doing.

It's a little worrying that you don't get the file descriptors back once the process ends; I guess because the kernel is killing the process quickly and it's not being cleaned up directly, but you'd expect the OS to be able to handle this case.

Anyways, these two posts seem to cover everything, and should solve the problem for you:

* http://superuser.com/a/838840
* http://superuser.com/a/443168

Definitely worth noting the current limits before changing them.

Let me know.  I know how frustrating rebooting is, and the goal of this project is to help you get more work done, not less :)

**Comment by clayne11** [comment]
Amazing thanks so much for your consistent support! I'm really hoping this solves my issue.

I have 1214 JS files excluding `node_modules` so that's very possibly the issue.

**Comment by clayne11** [comment]
So far so good - I upped those values and I haven't had any issue since! If it keeps working all day I'm going to close this. Also I'm closing https://github.com/gadicc/meteor-hmr/issues/62 because it's certainly the same issue.

**Comment by billyvg** [comment]
I'm actually having this issue as well, I changed my open file limits to 65536.

```
=> Starting gadicc:hot-build Accelerator (STq) on port 4002.


undefined:0

^
SyntaxError: Unexpected end of input
    at Object.parse (native)
    at WebSocket.<anonymous> (/Users/billy/.meteor/packages/gadicc_ecmascript-hot/.1.3.2-refactor.12.1spge2x++os+web.browser+web.cordova/plugin.compile-ecmascript-hot.os/npm/node_modules/meteor/gadicc_hot-build/node_modules/meteor-hotload-accelerator/lib/accelerator.js:89:22)
    at WebSocket.emit (events.js:98:17)
    at Receiver.self._receiver.ontext (/Users/billy/.meteor/packages/meteor-tool/.1.3.2_4.16uaxe7++os.osx.x86_64+web.browser+web.cordova/mt-os.osx.x86_64/dev_bundle/lib/node_modules/ws/lib/WebSocket.js:682:10)
    at Receiver.opcodes.1.finish (/Users/billy/.meteor/packages/meteor-tool/.1.3.2_4.16uaxe7++os.osx.x86_64+web.browser+web.cordova/mt-os.osx.x86_64/dev_bundle/lib/node_modules/ws/lib/Receiver.js:391:14)
    at /Users/billy/.meteor/packages/meteor-tool/.1.3.2_4.16uaxe7++os.osx.x86_64+web.browser+web.cordova/mt-os.osx.x86_64/dev_bundle/lib/node_modules/ws/lib/Receiver.js:372:33
    at Receiver.expectData (/Users/billy/.meteor/packages/meteor-tool/.1.3.2_4.16uaxe7++os.osx.x86_64+web.browser+web.cordova/mt-os.osx.x86_64/dev_bundle/lib/node_modules/ws/lib/Receiver.js:147:5)
    at Receiver.expectHandler (/Users/billy/.meteor/packages/meteor-tool/.1.3.2_4.16uaxe7++os.osx.x86_64+web.browser+web.cordova/mt-os.osx.x86_64/dev_bundle/lib/node_modules/ws/lib/Receiver.js:371:16)
    at Receiver.add (/Users/billy/.meteor/packages/meteor-tool/.1.3.2_4.16uaxe7++os.osx.x86_64+web.browser+web.cordova/mt-os.osx.x86_64/dev_bundle/lib/node_modules/ws/lib/Receiver.js:87:24)
    at Socket.firstHandler (/Users/billy/.meteor/packages/meteor-tool/.1.3.2_4.16uaxe7++os.osx.x86_64+web.browser+web.cordova/mt-os.osx.x86_64/dev_bundle/lib/node_modules/ws/lib/WebSocket.js:663:22)
```

**Comment by clayne11** [comment]
Did you update your `maxproc` limit as well? I upped mine to `2048`. I'm not sure if either of those are helping yet for sure as it's intermittent but I'll be sure to report back later.

**Comment by billyvg** [comment]
Please disregard, wiping `.meteor/local` seems to have helped.

**Comment by clayne11** [comment]
I'm closing this, I think the issue is solved!

Thanks again!

**Comment by gadicc** [comment]
@clayne11, great!  We've closed a few big issues this week.  A big thanks to you too, for reporting all the issues and sticking through the bug hunts despite the frustrations.  It's great to get rid of these issues before making a stable release.  I've added this to the the new troubleshooting doc, which I moved out of the README.

@billyvg, yeah, that's a different issue, I've seen it I think twice before, and another user reported it.  Let's track it in #71.
